File: sdks/go/challenges/ambiguous_logic.go
package challenges

import (
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/rand"
	"sort"

	xagentauth "github.com/dyshay/agentauth/sdks/go"
)

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type acceptableAnswer struct {
	Answer string  `json:"answer"` // hex-encoded result
	Score  float64 `json:"score"`  // 0-1, where 1.0 = primary
}

type scoredAnswerHash struct {
	AnswerHash string  `json:"answerHash"`
	Score      float64 `json:"score"`
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func xorBytesArr(data []byte, key int) []byte {
	result := make([]byte, len(data))
	for i := range data {
		result[i] = data[i] ^ byte(key&0xFF)
	}
	return result
}

func sortAscending(data []byte) []byte {
	result := make([]byte, len(data))
	copy(result, data)
	sort.Slice(result, func(i, j int) bool { return result[i] < result[j] })
	return result
}

func reverseBytes(data []byte) []byte {
	result := make([]byte, len(data))
	for i := range data {
		result[i] = data[len(data)-1-i]
	}
	return result
}

// ---------------------------------------------------------------------------
// Templates
// ---------------------------------------------------------------------------

type ambiguousTemplate struct {
	name     string
	generate func(data []byte, difficulty xagentauth.Difficulty) (string, []acceptableAnswer)
}

func luckyNumberGenerate(data []byte, difficulty xagentauth.Difficulty) (string, []acceptableAnswer) {
	byteCount := len(data)

	// Primary interpretation: 7 is "the" lucky number
	isLucky7 := byteCount == 7
	var primaryResult []byte
	if isLucky7 {
		primaryResult = xorBytesArr(data, 7)
	} else {
		primaryResult = xorBytesArr(data, 13)
	}

	alternatives := []acceptableAnswer{
		{Answer: hex.EncodeToString(primaryResult), Score: 1.0},
	}

	altLuckyNumbers := []int{3, 8, 13}
	altScores := map[int]float64{3: 0.6, 8: 0.6, 13: 0.7}

	if difficulty != xagentauth.DifficultyEasy {
		for _, lucky := range altLuckyNumbers {
			isLucky := byteCount == lucky
			var altResult []byte
			if isLucky {
				altResult = xorBytesArr(data, 7)
			} else {
				altResult = xorBytesArr(data, 13)
			}
			altHex := hex.EncodeToString(altResult)
			if altHex != alternatives[0].Answer {
				alternatives = append(alternatives, acceptableAnswer{Answer: altHex, Score: altScores[lucky]})
			}
		}
	}

	phrasings := []string{
		fmt.Sprintf("You are given %d bytes of data (base64-encoded).\nIf the number of bytes is a lucky number, XOR each byte with 7.\nOtherwise, XOR each byte with 13.\nReturn the hex-encoded result.", byteCount),
		fmt.Sprintf("The data below contains %d bytes.\nWhen the byte count is lucky, apply XOR 7 to every byte.\nWhen unlucky, apply XOR 13 instead.\nProvide your answer as a hex string.", byteCount),
	}

	return pickRandom(phrasings), alternatives
}

func famousConstantGenerate(data []byte, _ xagentauth.Difficulty) (string, []acceptableAnswer) {
	piResult := xorBytesArr(data, 31)
	eResult := xorBytesArr(data, 27)
	phiResult := xorBytesArr(data, 16)

	alternatives := []acceptableAnswer{
		{Answer: hex.EncodeToString(piResult), Score: 1.0},
		{Answer: hex.EncodeToString(eResult), Score: 0.8},
		{Answer: hex.EncodeToString(phiResult), Score: 0.6},
	}

	phrasings := []string{
		"XOR each byte of the provided data with the most famous mathematical constant's first two digits as an integer.\nReturn the hex-encoded result.",
		"Take the universally recognized mathematical constant, extract its first two digits as a whole number, and XOR every byte of the data with that number.\nProvide the hex-encoded output.",
	}

	return pickRandom(phrasings), alternatives
}

func bigSmallGenerate(data []byte, _ xagentauth.Difficulty) (string, []acceptableAnswer) {
	firstByte := data[0]

	var primary127 []byte
	if firstByte > 127 {
		primary127 = reverseBytes(data)
	} else {
		primary127 = sortAscending(data)
	}

	var alt100 []byte
	if firstByte > 100 {
		alt100 = reverseBytes(data)
	} else {
		alt100 = sortAscending(data)
	}

	var alt200 []byte
	if firstByte > 200 {
		alt200 = reverseBytes(data)
	} else {
		alt200 = sortAscending(data)
	}

	alternatives := []acceptableAnswer{
		{Answer: hex.EncodeToString(primary127), Score: 1.0},
	}

	alt100Hex := hex.EncodeToString(alt100)
	alt200Hex := hex.EncodeToString(alt200)

	if alt100Hex != alternatives[0].Answer {
		alternatives = append(alternatives, acceptableAnswer{Answer: alt100Hex, Score: 0.8})
	}
	if alt200Hex != alternatives[0].Answer && alt200Hex != alt100Hex {
		alternatives = append(alternatives, acceptableAnswer{Answer: alt200Hex, Score: 0.7})
	}

	phrasings := []string{
		"If the first byte of the data is big, reverse the entire byte array.\nOtherwise, sort all bytes in ascending order.\nReturn the hex-encoded result.",
		"Examine the first byte. If it is a big value, flip the array end-to-end.\nIf it is small, arrange bytes from lowest to highest.\nProvide the hex-encoded output.",
	}

	return pickRandom(phrasings), alternatives
}

var allAmbiguousTemplates = []ambiguousTemplate{
	{name: "lucky-number", generate: luckyNumberGenerate},
	{name: "famous-constant", generate: famousConstantGenerate},
	{name: "big-small", generate: bigSmallGenerate},
}

// ---------------------------------------------------------------------------
// Difficulty config
// ---------------------------------------------------------------------------

type ambiguousDiffConfig struct {
	dataSize      int
	templateCount int
}

var ambiguousDiffConfigs = map[xagentauth.Difficulty]ambiguousDiffConfig{
	xagentauth.DifficultyEasy:        {dataSize: 8, templateCount: 1},
	xagentauth.DifficultyMedium:      {dataSize: 16, templateCount: 1},
	xagentauth.DifficultyHard:        {dataSize: 32, templateCount: 2},
	xagentauth.DifficultyAdversarial: {dataSize: 64, templateCount: 3},
}

// ---------------------------------------------------------------------------
// AmbiguousLogicDriver
// ---------------------------------------------------------------------------

// AmbiguousLogicDriver implements challenges where instructions are
// deliberately ambiguous, testing how the agent handles uncertainty.
type AmbiguousLogicDriver struct{}

func (d *AmbiguousLogicDriver) Name() string { return "ambiguous-logic" }

func (d *AmbiguousLogicDriver) Dimensions() []xagentauth.ChallengeDimension {
	return []xagentauth.ChallengeDimension{xagentauth.DimensionReasoning, xagentauth.DimensionAmbiguity}
}

func (d *AmbiguousLogicDriver) EstimatedHumanTimeMs() int64 { return 45000 }
func (d *AmbiguousLogicDriver) EstimatedAITimeMs() int64    { return 1000 }

// Generate creates an ambiguous-logic challenge.
func (d *AmbiguousLogicDriver) Generate(difficulty xagentauth.Difficulty) (*xagentauth.ChallengePayload, string, error) {
	config := ambiguousDiffConfigs[difficulty]
	data := xagentauth.RandomBytes(config.dataSize)

	selectedTemplates := d.selectTemplates(config.templateCount)

	if len(selectedTemplates) == 1 {
		return d.generateSingle(selectedTemplates[0], data, difficulty)
	}
	return d.generateChained(selectedTemplates, data, difficulty)
}

// Verify checks if the submitted answer matches any acceptable answer.
// For ambiguous challenges, the answer hash corresponds to the primary answer,
// but we also accept alternatives (checking the context for scored answers).
func (d *AmbiguousLogicDriver) Verify(answerHash string, submitted string) (bool, error) {
	submittedHash := xagentauth.SHA256Hex([]byte(submitted))
	return xagentauth.TimingSafeEqual(answerHash, submittedHash), nil
}

func (d *AmbiguousLogicDriver) selectTemplates(count int) []ambiguousTemplate {
	// Shuffle
	shuffled := make([]ambiguousTemplate, len(allAmbiguousTemplates))
	copy(shuffled, allAmbiguousTemplates)
	rand.Shuffle(len(shuffled), func(i, j int) {
		shuffled[i], shuffled[j] = shuffled[j], shuffled[i]
	})
	if count > len(shuffled) {
		count = len(shuffled)
	}
	return shuffled[:count]
}

func (d *AmbiguousLogicDriver) generateSingle(
	tmpl ambiguousTemplate,
	data []byte,
	difficulty xagentauth.Difficulty,
) (*xagentauth.ChallengePayload, string, error) {
	instructions, acceptableAnswers := tmpl.generate(data, difficulty)

	scoredAnswers := d.hashAnswers(acceptableAnswers)
	answerHash := xagentauth.SHA256Hex([]byte(acceptableAnswers[0].Answer))

	contextMap := map[string]interface{}{
		"templateName":  tmpl.name,
		"primaryAnswer": acceptableAnswers[0].Answer,
		"scoredAnswers": scoredAnswers,
	}
	contextJSON, _ := json.Marshal(contextMap)

	payload := &xagentauth.ChallengePayload{
		Type:         "ambiguous-logic",
		Instructions: instructions,
		Data:         base64.StdEncoding.EncodeToString(data),
		Steps:        1,
		Context:      contextJSON,
	}

	return payload, answerHash, nil
}

func (d *AmbiguousLogicDriver) generateChained(
	templates []ambiguousTemplate,
	data []byte,
	difficulty xagentauth.Difficulty,
) (*xagentauth.ChallengePayload, string, error) {
	currentData := data
	var instructionParts []string
	var allAcceptable []acceptableAnswer

	for i, tmpl := range templates {
		instructions, answers := tmpl.generate(currentData, difficulty)
		instructionParts = append(instructionParts, fmt.Sprintf("--- Part %d ---\n%s", i+1, instructions))

		if i == 0 {
			allAcceptable = answers
		} else {
			var chained []acceptableAnswer
			for _, prev := range allAcceptable {
				prevData, _ := hex.DecodeString(prev.Answer)
				_, chainAnswers := tmpl.generate(prevData, difficulty)
				for _, ans := range chainAnswers {
					chained = append(chained, acceptableAnswer{
						Answer: ans.Answer,
						Score:  prev.Score * ans.Score,
					})
				}
			}
			allAcceptable = chained
		}

		// Use the primary answer as input for the next template
		currentData, _ = hex.DecodeString(allAcceptable[0].Answer)
	}

	// Deduplicate: keep highest-scoring version
	uniqueMap := make(map[string]float64)
	for _, ans := range allAcceptable {
		if existing, ok := uniqueMap[ans.Answer]; !ok || ans.Score > existing {
			uniqueMap[ans.Answer] = ans.Score
		}
	}

	var deduplicated []acceptableAnswer
	for answer, score := range uniqueMap {
		deduplicated = append(deduplicated, acceptableAnswer{Answer: answer, Score: score})
	}
	sort.Slice(deduplicated, func(i, j int) bool { return deduplicated[i].Score > deduplicated[j].Score })

	scoredAnswers := d.hashAnswers(deduplicated)
	answerHash := xagentauth.SHA256Hex([]byte(deduplicated[0].Answer))

	fullInstructions := "This is a multi-part ambiguous logic challenge.\nApply each part's transformation in order, using the output of the previous part as input for the next.\n\n"
	for i, part := range instructionParts {
		if i > 0 {
			fullInstructions += "\n\n"
		}
		fullInstructions += part
	}

	templateNames := make([]string, len(templates))
	for i, t := range templates {
		templateNames[i] = t.name
	}
	contextMap := map[string]interface{}{
		"templateNames": templateNames,
		"primaryAnswer": deduplicated[0].Answer,
		"scoredAnswers": scoredAnswers,
	}
	contextJSON, _ := json.Marshal(contextMap)

	payload := &xagentauth.ChallengePayload{
		Type:         "ambiguous-logic",
		Instructions: fullInstructions,
		Data:         base64.StdEncoding.EncodeToString(data),
		Steps:        len(templates),
		Context:      contextJSON,
	}

	return payload, answerHash, nil
}

func (d *AmbiguousLogicDriver) hashAnswers(answers []acceptableAnswer) []scoredAnswerHash {
	results := make([]scoredAnswerHash, len(answers))
	for i, ans := range answers {
		hash := xagentauth.SHA256Hex([]byte(ans.Answer))
		results[i] = scoredAnswerHash{AnswerHash: hash, Score: ans.Score}
	}
	return results
}

File: sdks/rust/src/challenges/ambiguous_logic.rs
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use rand::seq::SliceRandom;
use rand::Rng;

use crate::crypto::{sha256_hex, timing_safe_equal};
use crate::types::{ChallengeDimension, ChallengeDriver, ChallengePayload, Difficulty};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
struct AcceptableAnswer {
    answer: String, // hex-encoded result
    score: f64,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
struct ScoredAnswerHash {
    answer_hash: String,
    score: f64,
}

struct TemplateResult {
    instructions: String,
    acceptable_answers: Vec<AcceptableAnswer>,
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn xor_bytes(data: &[u8], key: u8) -> Vec<u8> {
    data.iter().map(|b| b ^ key).collect()
}

fn sort_ascending(data: &[u8]) -> Vec<u8> {
    let mut sorted = data.to_vec();
    sorted.sort();
    sorted
}

fn reverse_bytes(data: &[u8]) -> Vec<u8> {
    let mut result = data.to_vec();
    result.reverse();
    result
}

// ---------------------------------------------------------------------------
// Template 1: Lucky Number
// ---------------------------------------------------------------------------

fn lucky_number_template(data: &[u8], difficulty: &Difficulty) -> TemplateResult {
    let mut rng = rand::thread_rng();
    let byte_count = data.len();

    // Primary: 7 is "the" lucky number
    let is_lucky_7 = byte_count == 7;
    let primary_result = if is_lucky_7 {
        xor_bytes(data, 7)
    } else {
        xor_bytes(data, 13)
    };

    let mut alternatives = vec![AcceptableAnswer {
        answer: hex::encode(&primary_result),
        score: 1.0,
    }];

    // Alternative lucky numbers from various cultures
    if *difficulty != Difficulty::Easy {
        let alt_lucky = [(3usize, 0.6), (8, 0.6), (13, 0.7)];
        for &(lucky, score) in &alt_lucky {
            let is_lucky = byte_count == lucky;
            let alt_result = if is_lucky {
                xor_bytes(data, 7)
            } else {
                xor_bytes(data, 13)
            };
            let alt_hex = hex::encode(&alt_result);
            if alt_hex != alternatives[0].answer {
                alternatives.push(AcceptableAnswer {
                    answer: alt_hex,
                    score,
                });
            }
        }
    }

    let phrasings = [
        format!(
            "You are given {} bytes of data (base64-encoded).\n\
             If the number of bytes is a lucky number, XOR each byte with 7.\n\
             Otherwise, XOR each byte with 13.\n\
             Return the hex-encoded result.",
            byte_count
        ),
        format!(
            "The data below contains {} bytes.\n\
             When the byte count is lucky, apply XOR 7 to every byte.\n\
             When unlucky, apply XOR 13 instead.\n\
             Provide your answer as a hex string.",
            byte_count
        ),
    ];

    TemplateResult {
        instructions: phrasings[rng.gen_range(0..phrasings.len())].clone(),
        acceptable_answers: alternatives,
    }
}

// ---------------------------------------------------------------------------
// Template 2: Famous Constant
// ---------------------------------------------------------------------------

fn famous_constant_template(data: &[u8], _difficulty: &Difficulty) -> TemplateResult {
    let mut rng = rand::thread_rng();

    // Primary: pi -> "3.1" -> 31
    let pi_result = xor_bytes(data, 31);
    // Alternative: e -> "2.7" -> 27
    let e_result = xor_bytes(data, 27);
    // Alternative: phi -> "1.6" -> 16
    let phi_result = xor_bytes(data, 16);

    let alternatives = vec![
        AcceptableAnswer {
            answer: hex::encode(&pi_result),
            score: 1.0,
        },
        AcceptableAnswer {
            answer: hex::encode(&e_result),
            score: 0.8,
        },
        AcceptableAnswer {
            answer: hex::encode(&phi_result),
            score: 0.6,
        },
    ];

    let phrasings = ["XOR each byte of the provided data with the most famous mathematical constant's first two digits as an integer.\n\
         Return the hex-encoded result.".to_string(),
        "Take the universally recognized mathematical constant, extract its first two digits as a whole number, \
         and XOR every byte of the data with that number.\n\
         Provide the hex-encoded output.".to_string()];

    TemplateResult {
        instructions: phrasings[rng.gen_range(0..phrasings.len())].clone(),
        acceptable_answers: alternatives,
    }
}

// ---------------------------------------------------------------------------
// Template 3: Big/Small
// ---------------------------------------------------------------------------

fn big_small_template(data: &[u8], _difficulty: &Difficulty) -> TemplateResult {
    let mut rng = rand::thread_rng();
    let first_byte = data[0];

    // Primary: "big" means > 127
    let primary_127 = if first_byte > 127 {
        reverse_bytes(data)
    } else {
        sort_ascending(data)
    };

    // Alternative: "big" means > 100
    let alt_100 = if first_byte > 100 {
        reverse_bytes(data)
    } else {
        sort_ascending(data)
    };

    // Alternative: "big" means > 200
    let alt_200 = if first_byte > 200 {
        reverse_bytes(data)
    } else {
        sort_ascending(data)
    };

    let mut alternatives = vec![AcceptableAnswer {
        answer: hex::encode(&primary_127),
        score: 1.0,
    }];

    let alt_100_hex = hex::encode(&alt_100);
    let alt_200_hex = hex::encode(&alt_200);

    if alt_100_hex != alternatives[0].answer {
        alternatives.push(AcceptableAnswer {
            answer: alt_100_hex.clone(),
            score: 0.8,
        });
    }
    if alt_200_hex != alternatives[0].answer && alt_200_hex != alt_100_hex {
        alternatives.push(AcceptableAnswer {
            answer: alt_200_hex,
            score: 0.7,
        });
    }

    let phrasings = [
        "If the first byte of the data is big, reverse the entire byte array.\n\
         Otherwise, sort all bytes in ascending order.\n\
         Return the hex-encoded result."
            .to_string(),
        "Examine the first byte. If it is a big value, flip the array end-to-end.\n\
         If it is small, arrange bytes from lowest to highest.\n\
         Provide the hex-encoded output."
            .to_string(),
    ];

    TemplateResult {
        instructions: phrasings[rng.gen_range(0..phrasings.len())].clone(),
        acceptable_answers: alternatives,
    }
}

// ---------------------------------------------------------------------------
// Template dispatch
// ---------------------------------------------------------------------------

type TemplateFn = fn(&[u8], &Difficulty) -> TemplateResult;

const ALL_TEMPLATES: &[(&str, TemplateFn)] = &[
    ("lucky-number", lucky_number_template as TemplateFn),
    ("famous-constant", famous_constant_template as TemplateFn),
    ("big-small", big_small_template as TemplateFn),
];

// ---------------------------------------------------------------------------
// Difficulty config
// ---------------------------------------------------------------------------

fn difficulty_config(difficulty: &Difficulty) -> (usize, usize) {
    // (data_size, template_count)
    match difficulty {
        Difficulty::Easy => (8, 1),
        Difficulty::Medium => (16, 1),
        Difficulty::Hard => (32, 2),
        Difficulty::Adversarial => (64, 3),
    }
}

// ---------------------------------------------------------------------------
// Hash answers
// ---------------------------------------------------------------------------

fn hash_answers(answers: &[AcceptableAnswer]) -> Vec<ScoredAnswerHash> {
    answers
        .iter()
        .map(|a| ScoredAnswerHash {
            answer_hash: sha256_hex(a.answer.as_bytes()),
            score: a.score,
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Driver
// ---------------------------------------------------------------------------

pub struct AmbiguousLogicDriver;

impl AmbiguousLogicDriver {
    pub fn new() -> Self {
        Self
    }

    fn select_templates(&self, count: usize) -> Vec<(&str, TemplateFn)> {
        let mut rng = rand::thread_rng();
        let mut shuffled: Vec<_> = ALL_TEMPLATES.to_vec();
        shuffled.shuffle(&mut rng);
        shuffled.truncate(std::cmp::min(count, shuffled.len()));
        shuffled
    }

    fn generate_single(
        &self,
        template_name: &str,
        template_fn: TemplateFn,
        data: &[u8],
        difficulty: &Difficulty,
    ) -> Result<(ChallengePayload, String), String> {
        let result = template_fn(data, difficulty);
        let scored = hash_answers(&result.acceptable_answers);

        let primary_answer = &result.acceptable_answers[0].answer;
        let answer_hash = sha256_hex(primary_answer.as_bytes());

        let payload = ChallengePayload {
            challenge_type: "ambiguous-logic".into(),
            instructions: result.instructions,
            data: B64.encode(data),
            steps: 1,
            context: Some(serde_json::json!({
                "templateName": template_name,
                "primaryAnswer": primary_answer,
                "scoredAnswers": scored,
            })),
        };

        Ok((payload, answer_hash))
    }

    fn generate_chained(
        &self,
        templates: &[(&str, TemplateFn)],
        data: &[u8],
        difficulty: &Difficulty,
    ) -> Result<(ChallengePayload, String), String> {
        let mut current_data = data.to_vec();
        let mut instruction_parts = Vec::new();
        let mut all_acceptable: Vec<AcceptableAnswer> = Vec::new();

        for (i, (_name, func)) in templates.iter().enumerate() {
            let result = func(&current_data, difficulty);

            instruction_parts.push(format!("--- Part {} ---\n{}", i + 1, result.instructions));

            if i == 0 {
                all_acceptable = result.acceptable_answers;
            } else {
                let mut chained = Vec::new();
                for prev in &all_acceptable {
                    let prev_data = hex::decode(&prev.answer).unwrap_or_default();
                    let chain_result = func(&prev_data, difficulty);
                    for ans in &chain_result.acceptable_answers {
                        chained.push(AcceptableAnswer {
                            answer: ans.answer.clone(),
                            score: prev.score * ans.score,
                        });
                    }
                }
                all_acceptable = chained;
            }

            // Use primary for next template input
            current_data = hex::decode(&all_acceptable[0].answer).unwrap_or_default();
        }

        // Deduplicate by keeping highest score for each answer
        let mut unique: std::collections::HashMap<String, f64> = std::collections::HashMap::new();
        for ans in &all_acceptable {
            let entry = unique.entry(ans.answer.clone()).or_insert(0.0);
            if ans.score > *entry {
                *entry = ans.score;
            }
        }
        let mut deduplicated: Vec<AcceptableAnswer> = unique
            .into_iter()
            .map(|(answer, score)| AcceptableAnswer { answer, score })
            .collect();
        deduplicated.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap());

        let scored = hash_answers(&deduplicated);
        let primary_answer = &deduplicated[0].answer;
        let answer_hash = sha256_hex(primary_answer.as_bytes());

        let template_names: Vec<&str> = templates.iter().map(|(n, _)| *n).collect();
        let full_instructions = format!(
            "This is a multi-part ambiguous logic challenge.\n\
             Apply each part's transformation in order, using the output of the previous part as input for the next.\n\n\
             {}",
            instruction_parts.join("\n\n")
        );

        let payload = ChallengePayload {
            challenge_type: "ambiguous-logic".into(),
            instructions: full_instructions,
            data: B64.encode(data),
            steps: templates.len() as u32,
            context: Some(serde_json::json!({
                "templateNames": template_names,
                "primaryAnswer": primary_answer,
                "scoredAnswers": scored,
            })),
        };

        Ok((payload, answer_hash))
    }
}

impl Default for AmbiguousLogicDriver {
    fn default() -> Self {
        Self::new()
    }
}

impl ChallengeDriver for AmbiguousLogicDriver {
    fn name(&self) -> &str {
        "ambiguous-logic"
    }

    fn dimensions(&self) -> Vec<ChallengeDimension> {
        vec![ChallengeDimension::Reasoning, ChallengeDimension::Ambiguity]
    }

    fn estimated_human_time_ms(&self) -> u64 {
        45_000
    }

    fn estimated_ai_time_ms(&self) -> u64 {
        1_000
    }

    fn generate(&self, difficulty: &Difficulty) -> Result<(ChallengePayload, String), String> {
        let (data_size, template_count) = difficulty_config(difficulty);
        let data = crate::crypto::random_bytes(data_size);
        let selected = self.select_templates(template_count);

        if selected.len() == 1 {
            let (name, func) = selected[0];
            self.generate_single(name, func, &data, difficulty)
        } else {
            self.generate_chained(&selected, &data, difficulty)
        }
    }

    fn verify(&self, answer_hash: &str, submitted: &serde_json::Value) -> Result<bool, String> {
        let submitted_str = submitted
            .as_str()
            .ok_or_else(|| "Answer must be a string".to_string())?;
        let submitted_hash = sha256_hex(submitted_str.as_bytes());
        Ok(timing_safe_equal(answer_hash, &submitted_hash))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_payload_structure() {
        let driver = AmbiguousLogicDriver::new();
        let (payload, hash) = driver.generate(&Difficulty::Easy).unwrap();

        assert_eq!(payload.challenge_type, "ambiguous-logic");
        assert!(!payload.instructions.is_empty());
        assert!(!payload.data.is_empty());
        assert!(!hash.is_empty());
        assert_eq!(hash.len(), 64);
    }

    #[test]
    fn test_verify_correct_answer() {
        let driver = AmbiguousLogicDriver::new();
        let (payload, answer_hash) = driver.generate(&Difficulty::Easy).unwrap();

        let primary_answer = payload.context.as_ref().unwrap()["primaryAnswer"]
            .as_str()
            .unwrap()
            .to_string();
        let is_valid = driver
            .verify(&answer_hash, &serde_json::Value::String(primary_answer))
            .unwrap();
        assert!(is_valid);
    }

    #[test]
    fn test_verify_wrong_answer() {
        let driver = AmbiguousLogicDriver::new();
        let (_payload, answer_hash) = driver.generate(&Difficulty::Easy).unwrap();

        let is_valid = driver
            .verify(&answer_hash, &serde_json::Value::String("wrong".into()))
            .unwrap();
        assert!(!is_valid);
    }

    #[test]
    fn test_all_templates_produce_output() {
        let data = crate::crypto::random_bytes(16);
        for &(_name, func) in ALL_TEMPLATES {
            let result = func(&data, &Difficulty::Medium);
            assert!(!result.acceptable_answers.is_empty());
            assert!(!result.instructions.is_empty());
        }
    }
}
